internal/cmd/controller/gitops/reconciler/gitjob_controller.go

```diff
@@ -98,12 +99,42 @@ type GitJobReconciler struct {
 	Recorder        record.EventRecorder
 }
 
+// TypedResourceVersionUnchangedPredicate implements a update predicate to allow syncPeriod and RequeueAfter events
+type TypedResourceVersionUnchangedPredicate[T metav1.Object] struct {
+	predicate.TypedFuncs[T]
+}
+
+func isNil(arg any) bool {
+	if v := reflect.ValueOf(arg); !v.IsValid() || ((v.Kind() == reflect.Ptr ||
+		v.Kind() == reflect.Interface ||
+		v.Kind() == reflect.Slice ||
+		v.Kind() == reflect.Map ||
+		v.Kind() == reflect.Chan ||
+		v.Kind() == reflect.Func) && v.IsNil()) {
+		return true
+	}
+	return false
+}
+
+// Update implements default UpdateEvent filter for validating resource version change.
+func (TypedResourceVersionUnchangedPredicate[T]) Update(e event.TypedUpdateEvent[T]) bool {
+	if isNil(e.ObjectOld) {
+		return false
+	}
+	if isNil(e.ObjectNew) {
+		return false
+	}
+
+	return e.ObjectNew.GetResourceVersion() == e.ObjectOld.GetResourceVersion()
+}
+
 func (r *GitJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&v1alpha1.GitRepo{},
 			builder.WithPredicates(
 				// do not trigger for GitRepo status changes (except for commit changes)
 				predicate.Or(
+					TypedResourceVersionUnchangedPredicate[client.Object]{},
```

Comment:
Could you please elaborate on why this is needed? Since this is used with `predicate.Or`, wouldn't any update to the resources (which make the resource version change) make it return `true`, making the rest of predicates redundant?

Reply:
I added more information to the commit, as discussed this relates to https://github.com/kubernetes-sigs/controller-runtime/issues/2355